File: stacks/adt.py
MAX = 100

class Stack:
    def __init__(self):
        self.data = []
    
    def is_empty(self):
        return len(self.data) == 0

    def is_full(self):
        return len(self.data) == MAX

    def push(self, value):
        if self.is_full():
            print("Stack Overflow! Cannot push", value)
        else:
            self.data.append(value)
            print(f"{value} pushed to stack.")

    def pop(self):
        if self.is_empty():
            print("Stack Underflow! Cannot pop.")
        else:
            value = self.data.pop()
            print(f"Popped element: {value}")

    def display(self):
        if self.is_empty():
            print("Stack is empty.")
        else:
            print("Stack elements (top to bottom):", end=" ")
            for item in reversed(self.data):
                print(item, end=" ")
            print()


def main():
    stack = Stack()

    while True:
        print("1. Push (Insert element)")
        print("2. Pop (Remove top element)")
        print("3. Display (Show all elements)")
        print("4. Check if stack is empty")
        print("5. Check if stack is full")
        print("6. Exit")
        choice = input("Enter your choice (1–6): ")

        if choice == '1':
            value = int(input("Enter value to push: "))
            stack.push(value)
        elif choice == '2':
            stack.pop()
        elif choice == '3':
            stack.display()
        elif choice == '4':
            if stack.is_empty():
                print("Stack is EMPTY.")
            else:
                print("Stack is NOT empty.")
        elif choice == '5':
            if stack.is_full():
                print("Stack is FULL.")
            else:
                print("Stack is NOT full.")
        elif choice == '6':
            print("Exiting program.")
            break
        else:
            print("Invalid choice. Try again.")

if __name__ == "__main__":
    main()

File: stacks/adt.c
#include <stdio.h>
#include <stdlib.h>
#define MAX 5

typedef struct {
    int data[MAX];
    int top;
} Stack;

void init(Stack *s) {
    s->top = -1;
}

int isFull(Stack *s) {
    return s->top == MAX - 1;
}

int isEmpty(Stack *s) {
    return s->top == -1;
}

void push(Stack *s, int value) {
    if (isFull(s)) {
        printf("Stack Overflow! Cannot push %d\n", value);
        return;
    }
    s->data[++(s->top)] = value;
    printf("%d pushed to stack.\n", value);
}


int pop(Stack *s) {
    if (isEmpty(s)) {
        printf("Stack Underflow! Cannot pop.\n");
        return -1;
    }
    return s->data[(s->top)--];
}

void display(Stack *s) {
    if (isEmpty(s)) {
        printf("Stack is empty.\n");
        return;
    }
    printf("Stack elements (top to bottom): ");
    for (int i = s->top; i >= 0; i--) {
        printf("%d ", s->data[i]);
    }
    printf("\n");
}

int main() {
    Stack s;
    init(&s);
    int choice, value;

    while (1) {
        printf("1. Push (Insert element)\n");
        printf("2. Pop (Remove top element)\n");
        printf("3. Display (Show all elements)\n");
        printf("4. Check if stack is empty\n");
        printf("5. Check if stack is full\n");
        printf("6. Exit\n");
        printf("Enter your choice (1-6): ");
        scanf("%d", &choice);

        switch (choice) {
            case 1:
                printf("Enter value to push: ");
                scanf("%d", &value);
                push(&s, value);
                break;
            case 2:
                value = pop(&s);
                if (value != -1)
                    printf("Popped element: %d\n", value);
                break;
            case 3:
                display(&s);
                break;
            case 4:
                if (isEmpty(&s))
                    printf("Stack is EMPTY.\n");
                else
                    printf("Stack is NOT empty.\n");
                break;
            case 5:
                if (isFull(&s))
                    printf("Stack is FULL.\n");
                else
                    printf("Stack is NOT full.\n");
                break;
            case 6:
                printf("Exiting program.\n");
                exit(0);
            default:
                printf("Invalid choice. Try again.\n");
        }
    }

    return 0;
}
